hasktorch/src/Torch/Typed/Functional.hs

```diff
@@ -4318,6 +4372,62 @@ stack ::
   Tensor device dtype shape
 stack tensors = unsafePerformIO $ ATen.cast2 ATen.Managed.stack_ll tensors (natValI @dim :: Int)
 
+
+-- | Stack'
+
+type Stack' (dim :: Nat) (preShape :: [Nat]) (shape :: [Nat]) (n :: Nat) = (
+ KnownNat dim,
+ KnownNat n,
+ All KnownNat preShape,
+ All KnownNat shape,
+ ListLength shape ~ (ListLength preShape + 1),
+ n ~ Index shape dim
+ )
+
+-- | stack' - Untyped-esque stack that accepts a list of tensors 
+-- >>> t = ones :: CPUTensor 'D.Float '[]
+-- >>> t' = stack' @0 @'[] @'[1] [t]
+-- >>> :type t'
+-- t' :: Tensor '( 'D.CPU, 0) 'D.Float '[1]
+-- >>> dtype &&& shape &&& (\t'' -> D.asValue (toDynamic t'') :: [Float]) $ t'
+-- (Float,([1],[1.0]))
+-- >>> t = ones :: CPUTensor 'D.Float '[2,2]
+-- >>> t' = stack' @0 @'[2, 2] @'[1, 2, 2] [t]
+-- >>> :type t'
+-- t' :: Tensor '( 'D.CPU, 0) 'D.Float '[1, 2, 2]
+-- >>> dtype &&& shape &&& (\t'' -> D.asValue (toDynamic t'') :: [[[Float]]]) $ t'
+-- (Float,([1,2,2],[[[1.0,1.0],[1.0,1.0]]]))
+-- >>> t' = stack' @1 @'[2, 2] @'[2, 1, 2] [t]
+-- >>> :type t'
+-- t' :: Tensor '( 'D.CPU, 0) 'D.Float '[2, 1, 2]
+-- >>> dtype &&& shape &&& (\t'' -> D.asValue (toDynamic t'') :: [[[Float]]]) $ t'
+-- (Float,([2,1,2],[[[1.0,1.0]],[[1.0,1.0]]]))
+-- >>> t' = stack' @2 @'[2, 2] @'[2, 2, 1] [t]
+-- >>> :type t'
+-- t' :: Tensor '( 'D.CPU, 0) 'D.Float '[2, 2, 1]
+-- >>> dtype &&& shape &&& (\t'' -> D.asValue (toDynamic t'') :: [[[Float]]]) $ t'
+-- (Float,([2,2,1],[[[1.0],[1.0]],[[1.0],[1.0]]]))
+-- >>> t' = stack' @2 @'[2, 2] @'[2, 2, 3] [t, t, t]
+-- >>> :type t'
+-- t' :: Tensor '( 'D.CPU, 0) 'D.Float '[2, 2, 3]
+-- >>> dtype &&& shape &&& (\t'' -> D.asValue (toDynamic t'') :: [[[Float]]]) $ t'
+-- (Float,([2,2,3],[[[1.0,1.0,1.0],[1.0,1.0,1.0]],[[1.0,1.0,1.0],[1.0,1.0,1.0]]]))
+
+stack' ::
+  forall dim preShape shape dtype device n.
+  ( KnownNat dim,
+    Stack' dim preShape shape n
+  ) =>
+  -- | input list of tensors
+  [Tensor device dtype preShape] ->
+  -- | output
+  Tensor device dtype shape
+stack' tensors = case someNatVal (fromIntegral $ length tensors) of
```

Comment:
maybe I'm missing something here, but it seems like that the output shape is not at all computed from the input shape, is it? So, it's up to the user to know exactly how many elements the list has and do the shape computation by hand?
I understand that there can be desire to do this, but this is a very unsafe thing to do. A better way would be to use `vecStack` from below. You can convert your list first to a sized vector -- which can fail https://hackage.haskell.org/package/vector-sized-1.4.4/docs/Data-Vector-Generic-Sized.html#v:fromList -- and then use `vecStack` on it safely.